library/coqlib.ml

```diff
@@ -20,7 +20,7 @@ let coq = Libnames.coq_string (* "Stdlib" *)
 
 let init_dir = [ coq; "Init"]
 
-let jmeq_module_name = [coq;"Logic";"JMeq"]
+let jmeq_module_name = [coq;"Logic";"Base"; "Stdlib"; "Logic"; "JMeq"]
```

Comment:
I do not understand what such logical paths achieve. I guess it helps with compatibility to have the path suffix be `Stdlib.Logic.JMeq`. But why introduce this spurious `Base` module in the middle, which causes the path to become `Stdlib.Logic.Base.Stdlib.Logic.JMeq`? Unless I missed it, the rationale for `Base` is not in the CEP.

Reply:
It was somehow in the CEP: to clarify the structure of the Stdlib, the Logic directory had to be split into Base and Classical.
The repetition indeed is not ideal but enables to get a good backward compatibility with both forms of overqualified Require (i.e., `From Coq.Logic Require Foo` and `Require Coq.Logic.Foo`). At some point, if everyone switches to `From Stdlib Require Foo`, we'll be able to cleanup that repetition.

Reply:
Could you point me to the corresponding paragraph of the CEP? There are almost no occurrences of the words "base" and "classical" in the CEP, except in a few package names. If you are referring to the latter, I don't understand how package names are relevant to this change of logical paths.

Reply:
> we want to compile file A/B/C.v to path A.C while compiling A/D/E.v to A.E, I was not able to do it with recent dune

I don't know why this is needed, but the above hack seems super weird in the Coq model, as if you do `-Q dir A` things are not "uniform" anymore.

If you want `A/B/C.v` to be `A.C` why don't you put the file in `A/C.v` ?

We can add support for this in Dune, but I wonder what the user would even write in their `theory` definition file.

Reply:
> If you want `A/B/C.v` to be `A.C` why don't you put the file in `A/C.v` ?

Here the goal is to clarify the structure of the stdlib, that is the dependencies of `A/B` and `A/D` (or `Logic/Base` and `Logic/Classical` in stdlib) are very different.
So my understanding is that dune cannot do it, thanks for the answer.

Reply:
I still don't understand what you are trying to do, but for sure dune can certainly compile files in the way we determine. What I don't see is how departing from our classical structure, based on directories, clarifies the structure of the stdlib?

Reply:
As you can see, ideally filenames and paths would be the same thing but for historical/compatibility reason, here we need to depart a bit from that.

Reply:
I still don't understand what you are trying to do, what was the situation before and what it is now?

I seems to me that it may not be the best idea to stash these kind of changes on this PR; I read the reorganization part and indeed there are a lot of questions about it still open.

So I'd advise to defer this changes and do the split only here if you want to get this move forward quick.

As for the use case, I understand you want to split `Logic/*.v` into `Logic/Classical/*.v` and `Logic/Constructive/*.v`, and but still compile them as `Logic.*`. There is no problem doing this, but the question is why.

Unless you implement library / module aliases as in OCaml (or via more brute-force method), consumers of the `Logic.*` namespace still don't have a good way to select a subset of it.

In the current setup, where is let's say the file `Logic/Classical/Foo.vo` installed?

I agree that having aliases is a very desirable feature to refactor software. But note indeed how this is a purely tooling matter helping develop quicker, and aliases are totally unrelated to splitting the stdlib.

Note that this is totally unrelated to Dune, whatever the alias method Dune has no problem implementing it.

Reply:
> As for the use case, I understand you want to split `Logic/*.v` into `Logic/Classical/*.v` and `Logic/Constructive/*.v`, and but still compile them as `Logic.*`. There is no problem doing this

Great, then please do it

> Unless you implement library / module aliases as in OCaml (or via more brute-force method), consumers of the `Logic.*` namespace still don't have a good way to select a subset of it.

We don't need that

Reply:
> In the current setup, where is let's say the file Logic/Classical/Foo.vo installed?

?

Reply:
> Great, then please do it

But what happens if someone tries to do `Require Logic.Classical.Foo` , Coq will fail in this case.

> As you can see, ideally filenames and paths would be the same thing but for historical/compatibility reason, here we need to depart a bit from that.

Actually I cannot see that. Thinking a bit about it, I see that languages either use a flat namespace, or a hierarchical one. I am having trouble finding "hybrid" models.

The Coq model has important advantages, in particular about simplicity. Departing from it should be done with a careful specification of what the new model is.

Reply:
> > In the current setup, where is let's say the file Logic/Classical/Foo.vo installed?

In user-contrib/Stdlib/Logic/Foo.v

> But what happens if someone tries to do `Require Logic.Classical.Foo` , Coq will fail in this case.

Yes, that's the expected result since there is no Logic.Classical.Foo, only Logic.Foo

Reply:
So, this is now outside the scope of the current PR.